### PR TITLE
452 Add breadcrumb to the search result entries

### LINKED ISSUE #452: Add breadcrumb to the search result entries
When searching for documentation it would be nice to include the breadcrumb trail or url in the results page. Currently lots of the docs I've hit today all were for cloud but theres no way to know until you click into it. Having the breadcrumb/category/url in the result would make it easier at a glance to see what is what so I can zero into content that might help quicker.

Related: #451

### PR DESCRIPTION
**Pull request by AaronSadlerUK**
Added breadcrumbs to the search results with links to the relevant pages

resolves #452

### CODE REVIEW

File: OurUmbraco/Our/SearchResultExtensions.cs [hunk 1]
```diff
@@ -15,6 +18,25 @@ public static string GetTitle(this SearchResult result)
             return HttpContext.Current.Server.HtmlEncode(result.Fields.ContainsKey("nodeName") ? result["nodeName"] : string.Empty);
         }
 
+        public static IEnumerable<SearchResultBreadcrumbModel> GetBreadcrumbs(this SearchResult result)
+        {
+            var currentResult = new UmbracoHelper(UmbracoContext.Current).Content(result.Id) as IPublishedContent;
```
Comment (abjerner) [inline OurUmbraco/Our/SearchResultExtensions.cs:23]: I see a few issues with this line:

- Creating a new `UmbracoHelper` instance has a slight overhead, so initializing more than one per request should generally be avoided. In this line, an instance will be created for each search result. And the search seems to show up to 100 results, meaning a potential of 100 new instances. As our only use of `UmbracoHelper` is getting the IPublishedContent, we can do that directly via `UmbracoContext` instead - eg. by changing the line to:

    ```csharp
    var currentResult = UmbracoContext.Current.ContentCache.GetById(result.Id);
    ```

- Not all search results are based on content. For instance, forum topics are stored in the database, so when attempting to lookup a topic in the content cache, you will get `null` back - which will then result in an exception when trying to get the ancestors of `null` on the next line.

    So ideally, there should be a line line the one below between line 23 and line 24:

    ```csharp
    if (currentResult == null) return new List<SearchResultBreadcrumbModel>();
    ```

    Things don't always go according to plan, so you could also end up with inconsitencies between the content cache and the Examine index - eg. if a content item has been deleted, but still exists in the Examine index. The suggested null check above should also take care of that scenario 😉

File: OurUmbraco/Our/SearchResultExtensions.cs [hunk 2]
```diff
@@ -15,6 +18,25 @@ public static string GetTitle(this SearchResult result)
             return HttpContext.Current.Server.HtmlEncode(result.Fields.ContainsKey("nodeName") ? result["nodeName"] : string.Empty);
         }
 
+        public static IEnumerable<SearchResultBreadcrumbModel> GetBreadcrumbs(this SearchResult result)
```
Comment (abjerner) [inline OurUmbraco/Our/SearchResultExtensions.cs:21]: In the Razor view, you're calling `.ToList()` on the result of this method. Internally in the method it's already a list, so you might as well change the return type the return type to `List<SearchResultBreadcrumbModel>`. and then skip the call to `.ToList()` in the Razor might. That might help make the view look a bit more clean 😉

**Review by abjerner — CHANGES_REQUESTED** [review]
Hi @AaronSadlerUK and thanks for creating this PR 👍 

The search results could certainly do with a breadcrumb, so I had a look at your changes. Generally it seems to work really well, but I left some comments about a few of the lines that could be improved a bit.

As I'm mentioning in one of the comments, forum topics are not stored as content, so they won't be found when you try to do a lookup for them in the content cache. The proposed null check will make sure that users don't see a YSOD or error message, but then forum topics won't have a breadcrumb. So some additional changes should be made for that to happen - but unless you really want to look at tat as well now, I think we can skip that for now, and you or someone else can then have a look at that part later.

**Comment by AaronSadlerUK** [comment]
Thanks for the feedback @abjerner I will look at tweaking this :)

File: OurUmbraco/Our/SearchResultExtensions.cs (hunk 1 above)
Comment (AaronSadlerUK) [reply OurUmbraco/Our/SearchResultExtensions.cs]: I have made these changes

File: OurUmbraco/Our/SearchResultExtensions.cs (hunk 2 above)
Comment (AaronSadlerUK) [reply OurUmbraco/Our/SearchResultExtensions.cs]: I have made this change

**Comment by AaronSadlerUK** [comment]
I have added the breadcrumbs to the forums result:
image

**Comment by abjerner** [comment]
@AaronSadlerUK Great - I just had a quick look through the code, and it seems to have fixed the issues that I described. I'll see if I can give your PR another test tomorrow ;)

**Comment by AaronSadlerUK** [comment]
I have added the breadcrumbs to the documentation now =]

**Review by abjerner — APPROVED** [review]

**Comment by abjerner** [comment]
Great work 👍 

I had did another test, and everything seems to work like described, so I've approved your PR. There isn't really a review program for Our as there is for the CMS, so HQ will have to test and review this as well 😉 

<br /><br /><br /><br />

While testing, I did play a bit with the styling. Personally I think the breadcrumb could do with a slightly smaller text size, and a bit of bottom margin to create some distance to the text below. Eg. turning:

```css
.type-breadcrumb {
    color: #71888e;
    font-size: .95rem;
}
```
into:
```css
.type-breadcrumb {
    margin-bottom: 3px;
    color: #71888e;
    font-size: 0.85rem !important;
}
```

would make:

image

look like below instead:

image

Not really a big thing, so feel free to skip this part if you don't agree 😉

**Comment by AaronSadlerUK** [comment]
> ```css
> margin-bottom: 3px;
> ```

I agree, this is much better!

**Comment by AaronSadlerUK** [comment]
@abjerner I have added modified it to remove the last index from the documentation breadcrumbs.

I'm not sure why this wasn't working before:
```
if (strDirs.Last() == "index")
{
    strDirs.Remove(strDirs.Last());
}
```

**Comment by nul800sebastiaan** [comment]
Lovely work @AaronSadlerUK  and thanks for the thorough review @abjerner - great update! 👍 This will roll out to Our later today.

**Pull request merged**